### PR TITLE
[WIP] Refactor http client creation

### PR DESCRIPTION
**Pull request by wendigo**
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by wendigo** [comment]
Just rebased

**Comment by dominikzalewski** [comment]
Thanks for including me in the process of the review. As a new-joiner, I think this boosts my self-esteem.
I know that probably most of the team members are aware of what the problem is you are trying to solve, but for me without context, being only for a month or so in the project, it's totally abstract. I tried to browse the code changes, but it's way too extensive edits for me to somehow 'reverse-engineer' missing jira description from the code changes.

A good starting point for me would be a description of a business problem that these code changes address and possibly filling in the acceptance criteria for that change. I think this is in sync with @kokosing's question: is this a bug?

You can also ask a similar in latin: primum non nocere?
Or maybe, if it's the correct impression I get from this project: if it ain't broken, don't try to fix it?

File: client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java
Comment (hashhar) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java:1]: This is a no-go IMO. All user code which uses the JDBC driver expects to catch SQLException instead of RuntimeException.

File: client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoConnection.java [hunk 1]
```diff
@@ -119,7 +119,7 @@ public class TrinoConnection
         this.httpUri = uri.getHttpUri();
         uri.getSchema().ifPresent(schema::set);
         uri.getCatalog().ifPresent(catalog::set);
-        this.user = uri.getUser();
+        this.user = Optional.ofNullable(uri.getUser());
```
Comment (hashhar) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoConnection.java:122]: Does a `null` username actually work. The CLI defaults the username to OS username but if you use the REST API a `X-Trino-User` is always needed.

File: client/trino-client/src/main/java/io/trino/client/uri/HttpClientFactory.java
Comment (hashhar) [inline client/trino-client/src/main/java/io/trino/client/uri/HttpClientFactory.java:1]: This change can be extracted to it's own commit. i.e. converting existing code to use the factory instead.

Comment (hashhar) [reply client/trino-client/src/main/java/io/trino/client/uri/HttpClientFactory.java:1]: ConnectionProperties and related changes can be the next one.

File: client/trino-client/src/main/java/io/trino/client/uri/AbstractConnectionProperty.java [hunk 2]
```diff
@@ -162,26 +167,95 @@ public void validate(Properties properties)
 
     interface Converter<V, T>
     {
-        T convert(V value);
+        T decode(V value);
+
+        V encode(T value);
+    }
```
Comment (hashhar) [inline client/trino-client/src/main/java/io/trino/client/uri/AbstractConnectionProperty.java:173]: this interface change can also be extracted.

File: client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java
Comment (wendigo) [reply client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java:1]: Exception is converted on the call site

Comment (hashhar) [reply client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java:1]: Yes in the driver classes only. Note that `TrinoDriverUri` is public as well and part of the JDBC package.

File: client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoConnection.java (hunk 1 above)
Comment (wendigo) [reply client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoConnection.java:122]: yep, it works :)

File: client/trino-client/src/main/java/io/trino/client/uri/HttpClientFactory.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+                        redirectHandler, poller, knownTokenCache.create(), timeout);
+
+                builder.authenticator(authenticator);
+                builder.addInterceptor(authenticator);
+            }
+
+            uri.getDnsResolver().ifPresent(resolverClass -> builder.dns(instantiateDnsResolver(resolverClass, uri.getDnsResolverContext())::lookup));
+        }
+        catch (ClientException e) {
+            throw new RuntimeException(e.getMessage(), e);
```
Comment (nineinchnick) [inline client/trino-client/src/main/java/io/trino/client/uri/HttpClientFactory.java:154]: Why can't we keep using checked exceptions? Maybe add a code comment.

File: client/trino-client/src/main/java/io/trino/client/uri/PropertyName.java [hunk 4]
```diff
@@ -18,66 +18,78 @@
...
-    TIMEZONE("timezone"),
+    USER("user", "User name"),
+    PASSWORD("password", "User password"),
+    SESSION_USER("sessionUser", "Name of the user to impersonate"),
+    ROLES("roles", "Access roles that will be used to access catalogs"),
+    SOCKS_PROXY("socksProxy", "SOCKS proxy address"),
+    HTTP_PROXY("httpProxy", "HTTP proxy adress"),
+    APPLICATION_NAME_PREFIX("applicationNamePrefix", "Application name"),
+    DISABLE_COMPRESSION("disableCompression", "Disable compression"),
+    ASSUME_LITERAL_NAMES_IN_METADATA_CALLS_FOR_NON_CONFORMING_CLIENTS("assumeLiteralNamesInMetadataCallsForNonConformingClients", "tbd"),
```
Comment (nineinchnick) [inline client/trino-client/src/main/java/io/trino/client/uri/PropertyName.java:34]: Maybe `null` instead of `"tbd"`? If we don't have a good human-readable label for this, maybe code that would try to show this to users should throw exceptions.

File: client/trino-cli/src/main/java/io/trino/cli/QueryRunner.java [hunk 5]
```diff
@@ -15,54 +15,37 @@
...
-        OkHttpClient.Builder builder = new OkHttpClient.Builder();
-        try {
-            sslSetup = uri.getSetupSsl();
-            uri.setupClient(builder);
-        }
-        catch (SQLException e) {
-            throw new IllegalArgumentException(e);
-        }
-
-        setupTimeouts(builder, 30, SECONDS);
```
Comment (nineinchnick) [inline client/trino-cli/src/main/java/io/trino/cli/QueryRunner.java:60]: Do we want to keep a 30-second timeout as the default for the CLI? It defaults to 10 seconds in OkHTTP.

File: client/trino-cli/src/test/java/io/trino/cli/TestClientOptions.java [hunk 6]
```diff
@@ -162,15 +161,14 @@ public void testURLHostOnly()
 
     @Test
     public void testURLParams()
-            throws SQLException
     {
-        Console console = createConsole("trino://server.example:8080/my-catalog/my-schema?source=my-client");
```
Comment (nineinchnick) [inline client/trino-cli/src/test/java/io/trino/cli/TestClientOptions.java:167]: This whole test is for checking params, so removing this invalidates the test (maybe it duplicates another test?).

File: client/trino-client/src/main/java/io/trino/client/ClientSession.java [hunk 7]
```diff
@@ -303,6 +313,7 @@ public static final class Builder
         private Map<String, String> credentials = ImmutableMap.of();
         private String transactionId;
         private Duration clientRequestTimeout;
+        private DataSize targetResultSize = DataSize.of(1, MEGABYTE); // same as DEFAULT_TARGET_RESULT_SIZE
```
Comment (nineinchnick) [inline client/trino-client/src/main/java/io/trino/client/ClientSession.java:316]: Nit: adding `in the server` makes it more obvious why we don't use the same constant here.
```suggestion
        private DataSize targetResultSize = DataSize.of(1, MEGABYTE); // same as DEFAULT_TARGET_RESULT_SIZE in the server
```

File: client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java [hunk 8]
```diff
@@ -447,367 +150,431 @@ public URI getUri()
...
+    }
+
+    public Locale getLocale()
+    {
+        return resolve(LOCALE, Locale.getDefault());
+    }
+
+    public Duration getTimeout()
+    {
+        return resolve(TIMEOUT, Duration.valueOf("30s"));
```
Comment (nineinchnick) [inline client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java:430]: This changes the default (10s) that was used for JDBC. 30s was only for CLI.

File: client/trino-client/src/main/java/io/trino/client/uri/PropertyName.java (hunk 4 above)
Comment (nineinchnick) [reply client/trino-client/src/main/java/io/trino/client/uri/PropertyName.java:34]: If this is only used in `DriverPropertyInfo`, the docs say it can be null.

File: client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java [hunk 9]
```diff
@@ -447,367 +150,431 @@ public URI getUri()
...
-            String key = parts.get(0);
-            PropertyName name = PropertyName.findByKey(key).orElseThrow(() -> new SQLException(format("Unrecognized connection property '%s'", key)));
-            if (restrictedProperties.contains(name)) {
-                throw new RestrictedPropertyException(name, format("Connection property %s cannot be set in the URL", parts.get(0)));
+            return result;
+        }
+        catch (IllegalArgumentException e) {
+            String message = e.getMessage();
+            if (message.matches("Duplicate key \\[.*\\] found.")) {
+                throw new RuntimeException(format("Connection property %s is in the URL multiple times", extractPropertyName(message)));
```
Comment (nineinchnick) [inline client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java:511]: Can we catch this around the call to `put()` so we don't have to parse the error?

File: client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java [hunk 10]
```diff
@@ -818,65 +585,73 @@ private void initCatalogAndSchema()
...
+        List<RuntimeException> errors = violations.build();
+        if (errors.size() == 1) {
+            throw errors.get(0);
+        }
+        else if (!errors.isEmpty()) {
+            String multipleViolations = errors.stream()
+                    .map(RuntimeException::getMessage)
+                    .collect(joining("\n"));
+
+            throw new RuntimeException("Provided connection properties are invalid:\n" + multipleViolations);
```
Comment (nineinchnick) [inline client/trino-client/src/main/java/io/trino/client/uri/TrinoUri.java:654]: I'd join them without any prefix. Every error should form a correct English sentence, and we should terminate them with a dot. We shouldn't assume they'll be displayed to the user with newlines, some JDBC clients could mess up the formatting.

File: client/trino-client/src/test/java/io/trino/client/uri/TestTrinoUriBuilder.java [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        assertRoundtrip(HTTP_PROXY, HostAndPort.fromParts("proxy", 443), "proxy:443");
+        assertRoundtrip(APPLICATION_NAME_PREFIX, "prefix", "prefix");
+        assertRoundtrip(DISABLE_COMPRESSION, true, "true");
+        assertRoundtrip(ASSUME_LITERAL_NAMES_IN_METADATA_CALLS_FOR_NON_CONFORMING_CLIENTS, true, "true");
+        assertRoundtrip(ASSUME_LITERAL_UNDERSCORE_IN_METADATA_CALLS_FOR_NON_CONFORMING_CLIENTS, true, "true");
+        assertRoundtrip(SSL, true, "true");
+    }
+
+    @Test
+    public void testSetSimpleDependentValues()
```
Comment (nineinchnick) [inline client/trino-client/src/test/java/io/trino/client/uri/TestTrinoUriBuilder.java:66]: I'm not sure what this is testing, defaults?

File: client/trino-jdbc/src/main/java/io/trino/jdbc/NonRegisteringTrinoDriver.java [hunk 12]
```diff
@@ -37,13 +43,23 @@
...
         implements Driver, Closeable
 {
-    private final OkHttpClient httpClient = newHttpClient();
+    private static final String USER_AGENT = DRIVER_NAME + "/" + DRIVER_VERSION;
+    private final Dispatcher dispatcher;
+    private final ConnectionPool pool;
+
+    protected NonRegisteringTrinoDriver()
+    {
+        this.dispatcher = new Dispatcher();
```
Comment (nineinchnick) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/NonRegisteringTrinoDriver.java:52]: Shouldn't these be obtained from the client?

File: client/trino-jdbc/src/main/java/io/trino/jdbc/NonRegisteringTrinoDriver.java [hunk 13]
```diff
@@ -37,13 +43,23 @@
 public class NonRegisteringTrinoDriver
         implements Driver, Closeable
 {
-    private final OkHttpClient httpClient = newHttpClient();
+    private static final String USER_AGENT = DRIVER_NAME + "/" + DRIVER_VERSION;
+    private final Dispatcher dispatcher;
+    private final ConnectionPool pool;
+
+    protected NonRegisteringTrinoDriver()
```
Comment (nineinchnick) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/NonRegisteringTrinoDriver.java:50]: Why protected?

File: client/trino-jdbc/src/main/java/io/trino/jdbc/NonRegisteringTrinoDriver.java [hunk 14]
```diff
@@ -37,13 +43,23 @@
...
+    {
+        this.dispatcher = new Dispatcher();
+        this.pool = new ConnectionPool();
+    }
 
     @Override
     public void close()
+            throws IOException
```
Comment (nineinchnick) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/NonRegisteringTrinoDriver.java:58]: Why this was not needed before?

File: client/trino-jdbc/src/main/java/io/trino/jdbc/NonRegisteringTrinoDriver.java [hunk 15]
```diff
@@ -54,12 +70,16 @@ public Connection connect(String url, Properties info)
...
-        return new TrinoConnection(uri, instrumentClient(builder.build()));
+        try {
+            TrinoDriverUri uri = TrinoDriverUri.createDriverUri(url, info);
+            OkHttpClient.Builder httpClientBuilder = HttpClientFactory.toHttpClientBuilder(uri, USER_AGENT);
+            httpClientBuilder.connectionPool(pool);
+            httpClientBuilder.dispatcher(dispatcher);
+            return new TrinoConnection(uri, instrumentClient(httpClientBuilder.build()));
+        }
+        catch (RuntimeException e) {
+            throw new SQLException(e.getMessage(), e);
```
Comment (nineinchnick) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/NonRegisteringTrinoDriver.java:81]: Is it safe to assume all runtime exceptions are `SQLException`? Why can't the throw `SQLException` in the first place?

File: client/trino-jdbc/src/main/java/io/trino/jdbc/NonRegisteringTrinoDriver.java [hunk 16]
```diff
@@ -87,7 +107,22 @@ public boolean acceptsURL(String url)
     @Override
     public DriverPropertyInfo[] getPropertyInfo(String url, Properties info)
     {
-        return TrinoDriverUri.getPropertyInfo(url, info);
+        Properties properties = urlProperties(url, info);
+        return allProperties().stream()
+                .filter(property -> property.isValid(properties))
```
Comment (nineinchnick) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/NonRegisteringTrinoDriver.java:112]: Why is it ok to filter out invalid properties here, instead of throwing an exception? Add a code comment.

File: client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java [hunk 17]
```diff
@@ -33,15 +33,19 @@ public final class TrinoDriverUri
     private static final String JDBC_URL_START = JDBC_URL_PREFIX + "trino:";
 
     private TrinoDriverUri(String uri, Properties driverProperties)
-            throws SQLException
     {
         super(parseDriverUrl(uri), driverProperties);
     }
 
-    public static TrinoDriverUri create(String url, Properties properties)
+    public static TrinoDriverUri createDriverUri(String url, Properties properties)
```
Comment (nineinchnick) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java:40]: Nitpick level over 9000 - `createDriverUri` stutters with the class name.

File: client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java [hunk 18]
```diff
@@ -102,19 +103,19 @@ public class ClientOptions
...
-    @Option(names = "--krb5-service-principal-pattern", paramLabel = "<pattern>", defaultValue = "$${SERVICE}@$${HOST}", description = "Remote kerberos service principal pattern " + DEFAULT_VALUE)
+    @Option(names = "--krb5-service-principal-pattern", paramLabel = "<pattern>", description = "Remote kerberos service principal pattern " + DEFAULT_VALUE)
     public Optional<String> krb5ServicePrincipalPattern;
 
     @PropertyMapping(KERBEROS_REMOTE_SERVICE_NAME)
     @Option(names = "--krb5-remote-service-name", paramLabel = "<name>", description = "Remote peer's kerberos service name")
     public Optional<String> krb5RemoteServiceName;
 
     @PropertyMapping(KERBEROS_CONFIG_PATH)
-    @Option(names = "--krb5-config-path", paramLabel = "<path>", defaultValue = "/etc/krb5.conf", description = "Kerberos config file path " + DEFAULT_VALUE)
```
Comment (nineinchnick) [inline client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:113]: Won't this change the output of `trino --help`?

File: client/trino-client/src/main/java/io/trino/client/uri/PropertyName.java (hunk 4 above)
Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/uri/PropertyName.java:34]: I'll drop this change entirely.

File: client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java (hunk 17 above)
Comment (wendigo) [reply client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoDriverUri.java:40]: yes but clashes with the parent method that doesn't throw SQLException

File: client/trino-cli/src/main/java/io/trino/cli/QueryRunner.java (hunk 5 above)
Comment (wendigo) [reply client/trino-cli/src/main/java/io/trino/cli/QueryRunner.java:60]: Yes. I don't want to change defaults yet.

File: client/trino-cli/src/test/java/io/trino/cli/TestClientOptions.java (hunk 6 above)
Comment (wendigo) [reply client/trino-cli/src/test/java/io/trino/cli/TestClientOptions.java:167]: I reverted this change.

**Comment by mosabua** [comment]
Is this a duplicate of another PR for this or am I just misremembering? In any case .. I will leave this open and assume you will continue at some stage @wendigo

**Comment by mosabua** [comment]
👋 @wendigo I am leaving this open .. just fyi

**Comment by mosabua** [comment]
I am reopening since this seems like it is still something we are discussing and working on.
